src/lib/input/input-wrapper.ts

```diff
@@ -0,0 +1,289 @@
+import {
+  Component,
+  Input,
+  Directive,
+  AfterContentInit,
+  ContentChild,
+  SimpleChange,
+  ContentChildren,
+  ElementRef,
+  QueryList,
+  OnChanges,
+  ViewEncapsulation
+} from '@angular/core';
+import {MdError, coerceBooleanProperty} from '../core';
+import {NgModel} from '@angular/forms';
+
+
+// Invalid input type. Using one of these will throw an MdInputWrapperUnsupportedTypeError.
+const MD_INPUT_INVALID_INPUT_TYPE = [
+  'file',
+  'radio',
+  'checkbox',
+];
+
+
+let nextUniqueId = 0;
+
+
+export class MdInputWrapperInputElementCountError extends MdError {
+  constructor(count: number) {
+    super(`md-input-wrapper must contain exactly 1 input or textarea element. Found ${count}.`);
+  }
+}
+
+export class MdInputWrapperPlaceholderConflictError extends MdError {
+  constructor() {
+    super('Placeholder attribute and child element were both specified.');
+  }
+}
+
+export class MdInputWrapperUnsupportedTypeError extends MdError {
+  constructor(type: string) {
+    super(`Input type "${type}" isn't supported by md-input-wrapper.`);
+  }
+}
+
+export class MdInputWrapperDuplicatedHintError extends MdError {
+  constructor(align: string) {
+    super(`A hint was already declared for 'align="${align}"'.`);
+  }
+}
+
+/**
+ * The placeholder directive. The content can declare this to implement more
+ * complex placeholders.
+ */
+@Directive({
+  selector: 'md-placeholder'
+})
+export class MdPlaceholder {}
+
+
+/** The hint directive, used to tag content as hint labels (going under the input). */
+@Directive({
+  selector: 'md-hint',
+  host: {
+    '[class.md-right]': 'align == "end"',
+    '[class.md-hint]': 'true'
```

Comment:
switched to @HostBinding / @HostListener (since I realized that @HostListener actually does have a way to pass variables).